src/snowflake/cli/_plugins/nativeapp/entities/application_package.py

```diff
@@ -202,12 +216,15 @@ def warehouse(self) -> str:
         ) or to_identifier(self._workspace_ctx.default_warehouse)
 
     @property
-    def stage_fqn(self) -> str:
-        return f"{self.name}.{self._entity_model.stage}"
+    def scratch_stage_path(self) -> DefaultStagePathParts:
+        return DefaultStagePathParts(f"{self.name}.{self._entity_model.scratch_stage}")
 
-    @property
-    def scratch_stage_fqn(self) -> str:
-        return f"{self.name}.{self._entity_model.scratch_stage}"
+    @cached_property
+    def stage_path(self) -> DefaultStagePathParts:
+        stage_fqn = f"{self.name}.{self._entity_model.stage}"
+        subdir = self._entity_model.stage_subdirectory
+        full_path = f"{stage_fqn}/{subdir}" if subdir else stage_fqn
+        return DefaultStagePathParts(full_path)
```

Comment:
we could also accept different constructors in DefaultStagePathParts, so DefaultStagePathParts to undo the concat.